server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java

```diff
@@ -366,8 +394,17 @@ public AllocationStatus getLastAllocationStatus() {
      *
      * @return calculated delay in nanoseconds
      */
-    public long getRemainingDelay(final long nanoTimeNow, final Settings indexSettings) {
-        long delayTimeoutNanos = INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.get(indexSettings).nanos();
+    public long getRemainingDelay(
+        final long nanoTimeNow,
+        final Settings indexSettings,
+        final Map<String, SingleNodeShutdownMetadata> nodesShutdownMap
+    ) {
+        long delayTimeoutNanos = Optional.ofNullable(lastAllocatedNodeId)
+            .map(nodesShutdownMap::get)
+            .filter(shutdownMetadata -> SingleNodeShutdownMetadata.Type.RESTART.equals(shutdownMetadata.getType()))
+            .map(SingleNodeShutdownMetadata::getShardReallocationDelay)
+            .map(TimeValue::nanos)
+            .orElse(INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.get(indexSettings).nanos());
```

Comment:
> That seems wrong? I think we should take the greater of the shutdown delay and the delayed allocation delay.

I think it should be the explicitly set value in the node shutdown, regardless of the cluster-level setting for index delayed node left timeout. In that way we are doing exactly what the user asked (you asked me to wait 10 minutes, so I will wait 10 minutes regardless of the other setting) which is more expected. Why do you think it should be the max? That feels a little like action-at-a-distance when a user could not even be aware of the other cluster-wide setting.

Reply:
`index.unassigned.node_left.delayed_timeout` is an index level setting. The shutdown allocation delay is less likely to be controllable by a user (I believe cloud will simply use our default value) and I think of it as a "default value" for the index specific delay, allowing some extra (but never less) leeway to handle a planned restart.

If a user setup `index.unassigned.node_left.delayed_timeout=15m` for some indices, it seems odd that a planned restart would use a smaller value for the specific index/shards.

Also, if the shutdown API was called with `allocation_timeout=0`, it would be somewhat strange to disregard the index specific delay. That would effectively mean that we would be treating a planned restart as a worse event than a crash.

I think of restarts as something happening automatically, based on autoscaling, upgrades or other maintenance (though it could be an admin doing this, but without discussing with users). So if users (who should control the index level settings) set up their indices with higher delays than the shutdown uses, I think that higher delay should be used.

Reply:
I've changed this to use the max of index level delay and restart delay.